packages/salesforcedx-vscode-core/test/jest/commands/util/baseDeployRetrieve.test.ts

```diff
@@ -0,0 +1,123 @@
+import * as betaDeployRetrieveUtils from '../../../../src/commands/util/betaDeployRetrieve';
+import { WorkspaceContext } from '../../../../src/context';
+import * as sdrUtils from '../../../../src/services/sdr/componentSetUtils';
+
+jest.mock('../../../../src/commands/baseDeployRetrieve', () => {
+  return {
+    ...jest.requireActual('../../../../src/commands/baseDeployRetrieve'),
+    RetrieveExecutor: jest.fn()
+  };
+});
+
+describe('DeployRetrieveExecutor', () => {
+  const mockWorkspaceContext = {} as any;
+  const dummySourceConflictError = {
+    name: 'SourceConflictError',
+    message: '2 conflicts detected',
+    data: [
+      {
+        state: 'Conflict',
+        fullName: 'Test_Apex_Class_1',
+        type: 'ApexClass',
+        filePath:
+          '/Users/kenneth.lewis/scratchpad/TestProject-…ault/classes/Test_Apex_Class_1.cls-meta.xml'
+      },
+      {
+        state: 'Conflict',
+        fullName: 'Test_Apex_Class_1',
+        type: 'ApexClass',
+        filePath:
+          '/Users/kenneth.lewis/scratchpad/TestProject-…/main/default/classes/Test_Apex_Class_1.cls'
+      }
+    ]
+  };
+  let workspaceContextGetInstanceSpy: jest.SpyInstance;
+  let setApiVersionOnStub: jest.SpyInstance;
+  let formatExceptionStub: jest.SpyInstance;
+
+  class TestDeployRetrieveExecutor extends DeployRetrieveExecutor<{}> {
+    public error: any;
+    public handleSourceConflictErrorCalled: boolean = false;
+    private errorName: string | undefined;
+
+    protected doOperation(
+      components: ComponentSet,
+      token?: CancellationToken | undefined
+    ): Promise<(DeployResult | RetrieveResult) | undefined> {
+      if (this.errorName) {
+        const e = {} as any;
+        e.name = this.errorName;
+        throw e;
+      }
+      throw dummySourceConflictError;
+    }
+    protected postOperation(
+      result: (DeployResult | RetrieveResult) | undefined
+    ): Promise<void> {
+      return Promise.resolve(undefined);
+    }
+    protected handleSourceConflictError(e: any): void {
+      this.handleSourceConflictErrorCalled = true;
+      this.error = e;
+    }
+    constructor(s: string, t: string, errorName?: string) {
+      super(s, t);
+      this.errorName = errorName;
+    }
+
+    protected getComponents(
+      response: ContinueResponse<{}>
+    ): Promise<ComponentSet> {
+      return new Promise(resolve => resolve(new ComponentSet()));
+    }
+  }
+
+  beforeEach(() => {
+    workspaceContextGetInstanceSpy = jest
+      .spyOn(WorkspaceContext, 'getInstance')
+      .mockReturnValue(mockWorkspaceContext);
+    setApiVersionOnStub = jest
+      .spyOn(sdrUtils, 'setApiVersionOn')
+      .mockImplementation(jest.fn());
+    formatExceptionStub = jest
+      .spyOn(betaDeployRetrieveUtils, 'formatException')
+      .mockImplementation(jest.fn());
+  });
+
+  describe('run', () => {
+    it('should handle a SourceConflictError', async () => {
```

Comment:
Maybe a 3rd test for the absence of any error? Not wanting to repeat too much of what's covered in the vscode-integration test, but that's the 3rd option here right?

Reply:
Well, my intention here was to cover only the new functionality in unit tests, and maybe have a follow up WI to convert the rest of the integ tests.  Absence of any error would just be a successful deploy and testing those steps, which I think are covered pretty well in [the integ tests](https://github.com/forcedotcom/salesforcedx-vscode/blob/96eab0c65d44965e6286e32eee4063f7bba05d74/packages/salesforcedx-vscode-core/test/vscode-integration/commands/baseDeployRetrieve.test.ts#L256).